ci: split tests into chunks

khaneliman opened the pull request:
Similar goal of https://github.com/nix-community/home-manager/pull/6539 but applying it to the actual tests themselves for parallelization. Again, blocked by being able to change our checks ( I can't do that )

We have lots of tests and would like to add more. However, adding more testing coverage comes at the cost of a slower CI when we run them sequentially. This adds test outputs that are chunked however we'd like to tune for batch sizes. Allowing us to create a parallelized CI workflow.

Been testing in my fork and seems to cut time in half https://github.com/khaneliman/home-manager/actions/runs/16084828656

### Description

<!--

Please provide a brief description of your change.

-->

### Checklist

<!--

Please go through the following checklist before opening a non-WIP
pull-request.

Also make sure to read the guidelines found at

  https://nix-community.github.io/home-manager/#sec-guidelines

-->

- [ ] Change is backwards compatible.

- [ ] Code formatted with `nix fmt` or
    `nix-shell -p treefmt nixfmt-rfc-style deadnix keep-sorted --run treefmt`.

- [ ] Code tested through `nix-shell --pure tests -A run.all`
    or `nix build --reference-lock-file flake.lock ./tests#test-all` using Flakes.

- [ ] Test cases updated/added. See [example](https://github.com/nix-community/home-manager/commit/f3fbb50b68df20da47f9b0def5607857fcc0d021#diff-b61a6d542f9036550ba9c401c80f00ef).

- [ ] Commit messages are formatted like

    ```
    {component}: {description}

    {long description}
    ```

    See [CONTRIBUTING](https://nix-community.github.io/home-manager/#sec-commit-style) for more information and [recent commit messages](https://github.com/nix-community/home-manager/commits/master) for examples.

- If this PR adds a new module

  - [ ] Added myself as module maintainer. See [example](https://github.com/nix-community/home-manager/blob/a51598236f23c89e59ee77eb8e0614358b0e896c/modules/programs/lesspipe.nix#L11).

#### Maintainer CC

<!--
If you are updating a module, please @ people who are in its `meta.maintainers` list.
If in doubt, check `git blame` for whoever last touched something.
-->

khaneliman committed:
ci: split tests into chunks

We have lots of tests and would like to add more. However, adding more
testing coverage comes at the cost of a slower CI when we run them
sequentially. This adds test outputs that are chunked however we'd like
to tune for batch sizes. Allowing us to create a parallelized CI
workflow.

Signed-off-by: Austin Horstman <khaneliman12@gmail.com>

zowoq commented:
Have you looks at using buildbot for the tests? After the initial run I'd expect it to be quick as most tests would be cached. Darwin would be a bit bottlenecked so I'd suggest only testing aarch64-darwin and not x86_64-darwin.

khaneliman commented:
> Have you looks at using buildbot for the tests? After the initial run I'd expect it to be quick as most tests would be cached. Darwin would be a bit bottlenecked so I'd suggest only testing aarch64-darwin and not x86_64-darwin.

Yes, that's been something I have thought about but wasn't sure how to get set up in the checks here and everything there.

zowoq commented:
As there is a separate `flake.nix` for the tests I'll need to implement `flake_dir` in buildbot but I think this is all that would need to be done in home-manager. I can configure buildbot to just run on a branch without enabling it for the entire repo so can see it working without causing any disruption in other PRs.

```diff
diff --git a/buildbot-nix.toml b/buildbot-nix.toml
new file mode 100644
index 00000000..55433a03
--- /dev/null
+++ b/buildbot-nix.toml
@@ -0,0 +1,3 @@
+attribute = "packages"
+flake_dir = "./tests"
+lock_file = "./flake.lock"
diff --git a/tests/flake.nix b/tests/flake.nix
index 34f39d35..f5d183d0 100644
--- a/tests/flake.nix
+++ b/tests/flake.nix
@@ -20,7 +20,10 @@
   outputs =
     { nixpkgs, ... }:
     let
-      forAllSystems = nixpkgs.lib.genAttrs nixpkgs.lib.systems.flakeExposed;
+      forAllSystems = nixpkgs.lib.genAttrs [
+        "aarch64-darwin"
+        "x86_64-linux"
+      ];
     in
     {
       devShells = forAllSystems (
```

khaneliman commented:
> As there is a separate `flake.nix` for the tests I'll need to implement `flake_dir` in buildbot but I think this diff is all that would need to be done in home-manager. I can configure buildbot to just run on a branch without enabling it for the entire repo so can see it working without causing any disruption in other PRs. I should be able to have this done in the next day or two.
> 
> ```diff
> diff --git a/buildbot-nix.toml b/buildbot-nix.toml
> new file mode 100644
> index 00000000..55433a03
> --- /dev/null
> +++ b/buildbot-nix.toml
> @@ -0,0 +1,3 @@
> +attribute = "packages"
> +flake_dir = "./tests"
> +lock_file = "./flake.lock"
> diff --git a/tests/flake.nix b/tests/flake.nix
> index 34f39d35..f5d183d0 100644
> --- a/tests/flake.nix
> +++ b/tests/flake.nix
> @@ -20,7 +20,10 @@
>    outputs =
>      { nixpkgs, ... }:
>      let
> -      forAllSystems = nixpkgs.lib.genAttrs nixpkgs.lib.systems.flakeExposed;
> +      forAllSystems = nixpkgs.lib.genAttrs [
> +        "aarch64-darwin"
> +        "x86_64-linux"
> +      ];
>      in
>      {
>        devShells = forAllSystems (
> ```

I think in nixvim we ended up creating a separate flake output for `buildbot` itself so that we could keep the flake building on all platforms and only forward certain platforms to the ci.

zowoq commented:
The buildbot changes are deployed now if you'd like to move forward with that, note that the `nix-community-buildbot` label no longer works, instead the repo needs to be added to the list in the infra repo.

khaneliman commented:
> The buildbot changes are deployed now if you'd like to move forward with that, note that the `nix-community-buildbot` label no longer works, instead the repo needs to be added to the list in the infra repo.

Thanks, I'll try and do that within the next couple days.

khaneliman committed:
tests/flake: add buildbot output

Filters tests to specific platforms for buildbot.

Signed-off-by: Austin Horstman <khaneliman12@gmail.com>

khaneliman committed:
buildbot-nix.toml: add file

Support testing on buildbot CI

Signed-off-by: Austin Horstman <khaneliman12@gmail.com>

khaneliman committed:
tests: integration tests only run on linux

Using dependencies that require linux.

Signed-off-by: Austin Horstman <khaneliman12@gmail.com>

zowoq commented:
https://buildbot.nix-community.org/#/builders/4466/builds/19
https://buildbot.nix-community.org/#/builders/4466/builds/20

khaneliman committed:
tests: forward only test chunks to buildbot

We currently only need the tests covered by test-chunks in CI. We will
migrate more tests into the buildbot CI, as we are able to flesh them
out more.

Signed-off-by: Austin Horstman <khaneliman12@gmail.com>

khaneliman committed:
ci: don't duplicate test runs on github

We have moved towards running more tests on buildbot for better
performance. Don't duplicate efforts on github actions.

Signed-off-by: Austin Horstman <khaneliman12@gmail.com>

zowoq commented:
https://buildbot.nix-community.org/#/builders/4466/builds/22

khaneliman commented:
> https://buildbot.nix-community.org/#/builders/4466/builds/22

Anything else we need to configure to get the buildbot status hooked up here? Or is that on your side once we get it in the default branch?

khaneliman committed:
ci: re-enable home manager install and uninstall tests on darwin

Signed-off-by: Austin Horstman <khaneliman12@gmail.com>

khaneliman commented:
> Sorry this has been awkward to test, I had a look at trying to fix it but I haven't found the cause yet.

Np, took a break from it since I was on vacation for a bit. 

> No, nothing on this side. I've deployed [nix-community/infra#1904](https://github.com/nix-community/infra/pull/1904) so if you rebase this PR buildbot should check it.

https://buildbot.nix-community.org/#/builders/4466/builds/23 sweet, looks like it works now

zowoq commented:
There was a brief issue with the buildbot webhook not working, seems it didn't like switching from the manual branch mode to github PR mode but it should be fine now.

khaneliman commented:
> There was a brief issue with the buildbot webhook not working, seems it didn't like switching from the manual branch mode to github PR mode but it should be fine now.

Appreciate you getting this setup, thanks for the help!
